### ISSUE TITLE
Possibility to remove Events from EventEmitter

### ISSUE THREAD

**Issue by tomhatzer** [caffeinery/coffea#79]
Hi guys!

I'm still working with coffea and I really like it :)

As I can see, there are 3 events that get created by the ".on()" method.

eg.

```javascript
function parseCmd(event) {
   
}
client.on('command', parseCmd);
```

Normally I could call 
```javascript
client.removeListener('command', parseCmd);
```
and the event would be gone, but not here. If I do execute the removeListener, nothing happens and the events stay where they are. Nothing gets removed.

Do you have an idea how I can remove the events properly?

Thanks!

Tom

**Comment by omnidan** [comment]
Thanks for finding this, @tomhatzer 

We use [eventemitter3](https://github.com/primus/eventemitter3) which should be 100% compatible with the [node event api](http://nodejs.org/api/events.html).

Can you post a gist with example code so I can easily reproduce the issue? :grin:

**Comment by tomhatzer** [comment]
There are 2 problems from my point of view :)

In file lib/plugins/command.js on line 16-24 you build a custom object that includes a few different variables together with the callback function:
```
            utils.emit(irc, event.network, 'command', {
                "channel": event.channel,
                "user": event.user,
                "message": event.text,
                "cmd": cmd,
                "args": message,
                "isAction": event.isAction,
                "tags": event.tags ? event.tags : []
            });
```

In file lib/utils.js on line 73-75 3 events get created with a custom event_obj:
```
    var ret1 = instance.emit(event, err, event_obj);
    var ret2 = instance.emit(network + ":" + event, err, event_obj);
    var ret3 = instance.emit('event', event, err, event_obj);
```

Based on those 2 parts of code, it's impossible to remove binds for commands, because I can only guess what network it is bound to and what the contents of event_obj are.

I helped myself by creating a wrapper class for the internal nodejs EventEmitter that does nothing more than bind itself to message and command events and lets plugins bind to the Events from the wrapper class.

Long story short, my wrapper for this: https://gist.github.com/tomhatzer/73b24404195afe79d97d

Example code for bind/unbind:
```
var cmdHelper = require("CommandHelper.js");
var cb = function(event) {
    console.log(event);
};
cmdHelper.on('command', cb);
cmdHelper.removeListener('command', cb);
```

**Comment by omnidan** [comment]
@tomhatzer Glad you figured out a workaround :grin: I don't understand how dispatching the event three times makes it impossible to remove listeners.

If you listen to `on('command')`, you can remove the listener for that, it's totally separate from the `network:command` event. It's like listening to `on('command')` vs `on('connect')` - you can still remove the listener for either event.

**Comment by tomhatzer** [comment]
@omnidan :D The listeners for the internal EventEmitter need to be removed using .removeListener('command', callback); Doing this would work usually, but not in that case because I don't have the exact callback function that is used by eventemitter3. Also I don't have the exact context object for calling the eventemitter3 removeListener, which makes it impossible for me to remove the listeners for the 3 binds. There's always 1 variable thats missing or wrong (especially when working with multiple irc servers).

I created a plugin system that gives me the possibility to reload every plugin file on the fly (by removing the require cache entry for that file and loading it again). Which works now thanks to the CommandHelper class :D 

Everything else works as intended :)

**Comment by omnidan** [comment]
@tomhatzer how do you not have the exact callback function?

```
function callback (event) {
  event.reply('do something')
}

client.on('command', callback)
client.removeListener(callback)
```

That should work fine. You don't bind it three times, I only *emit* three times, and if you don't listen to `network:` or `event` it's not even emitted by the eventemitter.

**Comment by tomhatzer** [comment]
@omnidan I... I'm sorry, I mixed things up with the emits. :(

Hmmm, you all the removeListener with only 1 parameter.
I always called it with client.removeListener('command', callback);

**Comment by omnidan** [comment]
@tomhatzer Sorry, that was my mistake! You do have to [call it with the event name](https://nodejs.org/api/events.html#events_emitter_removelistener_event_listener), but it should still work fine like this:

```
function callback (event) {
  event.reply('do something')
}

client.on('command', callback)
client.removeListener('command', callback)
```

You won't have to unsubscribe from all three events (`command`, `network:command` and `event`), unless you subscribed to more than one of these.
